### PR TITLE
Add QgsAbstractGeometry::compareTo( QgsAbstractGeometry*)

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows for comparison of geometry objects, e.g. to allow for stable
sorting of them.

Ported from the GEOS equivalent method, but with addition of support
for M values and curved geometry types

### CODE REVIEW

File: src/core/geometry/qgsabstractgeometry.cpp [hunk 1]
```diff
@@ -40,6 +40,39 @@ QgsAbstractGeometry &QgsAbstractGeometry::operator=( const QgsAbstractGeometry &
...
+  // compare to self
+  if ( this == other )
+  {
+    return 0;
+  }
+
+  if ( sortIndex() != other->sortIndex() )
+  {
+    //different geometry types
+    int diff = sortIndex() - other->sortIndex();
```
Comment (pblottiere) [inline src/core/geometry/qgsabstractgeometry.cpp:54]:
```suggestion
    const int diff = sortIndex() - other->sortIndex();
```

File: src/core/geometry/qgscompoundcurve.cpp [hunk 2]
```diff
@@ -67,6 +67,37 @@ QgsCompoundCurve *QgsCompoundCurve::createEmptyWithSameType() const
...
+  if ( !otherCurve )
+    return -1;
+
+  int i = 0;
+  int j = 0;
+  while ( i < mCurves.size() && j < otherCurve->mCurves.size() )
+  {
+    const QgsAbstractGeometry *aGeom = mCurves[i];
+    const QgsAbstractGeometry *bGeom = otherCurve->mCurves[j];
+    int comparison = aGeom->compareTo( bGeom );
```
Comment (pblottiere) [inline src/core/geometry/qgscompoundcurve.cpp:82]:
```suggestion
    const int comparison = aGeom->compareTo( bGeom );
```

File: src/core/geometry/qgsgeometrycollection.cpp [hunk 3]
```diff
@@ -1045,3 +1045,34 @@ QgsAbstractGeometry *QgsGeometryCollection::childGeometry( int index ) const
...
+  if ( !otherCollection )
+    return -1;
+
+  int i = 0;
+  int j = 0;
+  while ( i < mGeometries.size() && j < otherCollection->mGeometries.size() )
+  {
+    const QgsAbstractGeometry *aGeom = mGeometries[i];
+    const QgsAbstractGeometry *bGeom = otherCollection->mGeometries[j];
+    int comparison = aGeom->compareTo( bGeom );
```
Comment (pblottiere) [inline src/core/geometry/qgsgeometrycollection.cpp:1061]:
```suggestion
    const int comparison = aGeom->compareTo( bGeom );
```

File: src/core/geometry/qgsabstractgeometry.h [hunk 4]
```diff
@@ -1026,6 +1033,22 @@ class CORE_EXPORT QgsAbstractGeometry
...
+     * Returns the sort index for the geometry, used in the compareTo() method to compare
+     * geometries of different types.
+     *
+     * \since QGIS 3.20
+     */
+    int sortIndex() const;
+
+    /**
+     * Compares to an \a other geometry of the same class, and returns a integer
+     * for sorting of the two geometries.
```
Comment (lbartoletti) [inline src/core/geometry/qgsabstractgeometry.h:1046]: Can you add the information on the returned result, I mean what corresponds to -1, 1, 0?

Comment (lbartoletti) [reply src/core/geometry/qgsabstractgeometry.h:1046]: BTW, shouldn't this be an enum?

File: tests/src/core/testqgsgeometry.cpp [hunk 5]
```diff
@@ -17180,6 +17183,148 @@ void TestQgsGeometry::createEmptyWithSameType()
   testCreateEmptyWithSameType<QgsTriangle>();
 }
 
+void TestQgsGeometry::compareTo_data()
+{
+  QTest::addColumn<QString>( "geom1" );
+  QTest::addColumn<QString>( "geom2" );
+  QTest::addColumn<int>( "expected" );
+
+  QTest::newRow( "point vs multipoint" ) << QStringLiteral( "POINT( 1 2 )" ) << QStringLiteral( "MULTIPOINT((1 2))" ) << -1;
```
Comment (lbartoletti) [inline tests/src/core/testqgsgeometry.cpp:17192]: What's the result with EMPTY type?

File: src/core/geometry/qgsabstractgeometry.h (hunk 4 above)
Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.h:1046]: > Can you add the information on the returned result, I mean what corresponds to -1, 1, 0?

I've added a note stating that it's an internal detail only and subject to change between qgis versions :+1:

Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.h:1046]: > BTW, shouldn't this be an enum?

No, c++ sorting algorithms always use a int only.

File: tests/src/core/testqgsgeometry.cpp (hunk 5 above)
Comment (nyalldawson) [reply tests/src/core/testqgsgeometry.cpp:17192]: I've added tests for these now

**Pull request merged**